### DISCUSSION TITLE
XML Support

### DISCUSSION THREAD

**Discussion by deluan** [opensubsonic/open-subsonic-api#69] [General]
Hey folks, I've been away for a while, and there's quite a lot to catch up here. Not sure if you discussed this before, but should we support XML in OpenSubsonic, or just JSON? I'm asking because I don't see how XML payloads should be represented for example for the `genres` attribute. This is what I'm currently getting:
```xml
<subsonic-response xmlns="http://subsonic.org/restapi" status="ok" version="1.8.0" type="navidrome" serverVersion="v0.0.0" openSubsonic="true">
  <album id="1" name="N" userRating="0" musicBrainzId="1234567890">
    <genres name="Rock"></genres>
    <genres name="Pop"></genres>
    <song id="1" isDir="true" title="title" album="album" artist="artist" track="1" year="1985" genre="Rock" coverArt="1" size="8421341" contentType="audio/flac" suffix="flac" starred="2022-11-21T17:21:00Z" transcodedContentType="audio/mpeg" transcodedSuffix="mp3" duration="146" bitRate="320" isVideo="false" bpm="127" comment="a comment">
      <genres name="Rock"></genres>
      <genres name="Pop"></genres>
    </song>
  </album>
</subsonic-response>
```

I'm pretty sure this is not what we want. I'm not sure if we can have duplicated attributes, like:
```XML
  <album id="1" name="N" userRating="0" musicBrainzId="1234567890" genre="Rock" genre="Pop"/>
```

So, what should we do? Was this discussed at all? Couldn't find anything in a quick search.

**Comment by dweymouth** [comment]
At some point @sentriz mentioned that we should follow the way Go serializes arrays in XML, which is exactly what this is. And Supersonic is actually using XML right now, since I started from forking an unmaintained API client library that used XML. I am not opposed to switching that library to JSON - just pointing out that this is what Gonic and LMS are already doing for XML and it works fine.

**Reply by dweymouth** [reply]
But if we deprecate XML for OpenSubsonic, we just need to wait until after a Supersonic release where I've switched its underlying library over to Json :) Not sure if Symfonium uses XML or Json, but as of now, we are the only 2 OpenSubsonic clients that I know of.

**Comment by sentriz** [comment]
that looks right to me, and it also looks like our existing support for array-like things in the subsonic XML protocol. for example the search3 response has a list of artist, song, track, etc - and it looks the same as that. this is what i have for gonic https://go.dev/play/p/A-4iicX2fP1


i would not be in favour of ditching XML personally because it becomes a pain to keep internal types that marshal to JSON all of the time but marshal to XML only some of the time (assuming we would support fancier types)

and I still use XML only clients like DSub and jamstash occasionally

**Comment by deluan** [comment]
Ah, thanks for the clarification folks. And yeah @sentriz, I'm also not in favour of ditching XML, as DSub is may main music player at the moment, and I want to keep supporting it.

**Comment by Tolriq** [comment]
There's no ditching of XML since the goal of OS is compatibility :)
If we eventually remove XML for new endpoints it will no have any impact on older clients anyway.
